### PR TITLE
`Communication`: Allow tutors to propose FAQ

### PR DESCRIPTION
**Pull request by cremertim**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Tutors are mainly answering questions of the students, so they know what is feaquently asked. Consequently, they should be able to propose FAQs

### Description

I allowed tutors of the course to propose FAQ's

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Tutor
- Course with FAQ enabled

1. Log in to Artemis as Tutor
2. Navigate to Course Management
3. Select FAQ
4. Propse atleast 2 FAQ's
5. Check, that the proposed FAQ is **not** shown in the Course Overview 
6. Check, that a tutor is not able to edit a accepted FAQ

1. Log in as Instructor
2. Navigate to Course Management
3. See newly proposed FAQs
4. Accept / Reject FAQ
5. Go to Course Overview and verify, only accepted FAQ is shown

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
|faq-component.ts | 96.66% | ✅                           |
|faq-update-component.ts| 96.05% | ✅              |
|faq-service.ts| 100% | ✅              |
|course-faq-component.ts| 100% | ✅              |

### Screenshots

image
Tutor View

image
Instructor View

## Summary by CodeRabbit

- **New Features**
	- Introduced new methods for retrieving FAQs based on course ID and status.
	- Enhanced FAQ management with role-based access controls for creating, updating, and displaying FAQs.
	- Added new labels for FAQ actions in multiple languages.
	- Improved conditional rendering for FAQ actions based on user roles.

- **Bug Fixes**
	- Improved handling of FAQ states during creation and updates.

- **Documentation**
	- Updated test cases to reflect changes in FAQ state management and retrieval methods.

## Release Notes

- **New Features**
	- Enhanced FAQ management with new methods for retrieving FAQs based on course ID and status.
	- New endpoint for fetching FAQs by status, accessible to students.
	- Conditional rendering of FAQ links updated to allow tutor access.
	- Localization improvements for FAQ messages in English and German.

- **Bug Fixes**
	- Authorization checks refined for FAQ operations based on user roles.

- **Documentation**
	- Updated user feedback messages for FAQ actions, providing more clarity and context.

### CODE REVIEW

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (ole-ve) [inline src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:188]: Was this unintentionally there or is there a specific reason this has been removed now?

File: src/main/webapp/i18n/de/faq.json
Comment (ole-ve) [inline src/main/webapp/i18n/de/faq.json:13]: "Das" oder "die" Nutella?

**Review by ole-ve — APPROVED** [review]
Code looks good, added two minors.

File: src/main/webapp/i18n/de/faq.json
Comment (cremertim) [reply src/main/webapp/i18n/de/faq.json:13]: Good question. I Actually dont care whether we use Das or Der. What do you prefer?

Comment (florian-glombik) [reply src/main/webapp/i18n/de/faq.json:13]: image

In case my short google research is correct I think that the sentence is in nominativ and therefore `Die` would be correct?
However, I might be mistaken and have no strong opinion.

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:121]: Please use the `@throws` tag and also add a description above the params

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:123]: Something like this would be clearer. Otherwise I expect it to just check for instructor status
```suggestion
    private void checkPriviledgeForAcceptedElseThrow(Faq faq, Long courseId) {
```

Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:123]: Should tutors be allowed to change an accepted faq to unaccepted again? Otherwise you probably have to verify this here as well

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:193]:
```suggestion
    public ResponseEntity<Set<FaqDTO>> getAllFaqsForCourseByStatus(@PathVariable Long courseId, @PathVariable String faqState) {
```

Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:193]: Can't you directly use the enum as parameter here? I think we do this in other endpoints as well

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:217]: Since you already removed the other verification:
```suggestion
    @EnforceAtLeastStudentInCourse
```

File: src/main/webapp/app/faq/faq.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/faq/faq.component.ts:57]: We want to make our icons `readonly`

File: src/main/webapp/i18n/de/faq.json
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/faq.json:15]: Does the id really help an instructor?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/webapp/i18n/de/faq.json
Comment (cremertim) [reply src/main/webapp/i18n/de/faq.json:15]: I think this is personal preference. I prefer to get feedback about what i edited instead of an generic message. Otherwise, you might be confused if you accidently clicked on the wrong FAQ / you can double check the faq in the overview easier if you know the id

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (cremertim) [reply src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:123]: Good catch, thank you. I adapted the update method

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (florian-glombik) [inline src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:78]: Could you please have another look at the permissions in this Resource? 
This PR description says that editors+ should be allowed to propose FAQs but here it seems like tutors+ are allowed to even `create` FAQs? 

Do I misunderstand something here?

Comment (florian-glombik) [reply src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:78]: I do now see that `checkPriviledgeForAcceptedElseThrow` seems to cover the permission check. Still the inconsistency with tutor vs editor exists

Is there a particular reason why this is handled in one endpoint? I think it would be cleaner to have a separate endpoint for proposing a faq 😅

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (florian-glombik) [inline src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:111]: Why do we not directly check for `@EnforeAtLeastInstructorInCourse`?

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (florian-glombik) [inline src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:210]: Why dont we have the course check here? 

```suggestion
    @EnforceAtLeastStudentInCourse
```

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (florian-glombik) [reply src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:78]: Ah well, just the motivation and description seems to differ from each other, so I guess the tutor change is intended 😅 
So it would be good if the description could be updated

File: src/main/webapp/i18n/de/faq.json
Comment (florian-glombik) [reply src/main/webapp/i18n/de/faq.json:15]: I think the title would be more helpful here 😅

**Review by florian-glombik — COMMENTED** [review]
I have some questions regarding the server side resource, the client side code looks good to me except for the success message where I think the FAQ title would be more useful

Could you double check / answer these questions / comments?

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (cremertim) [reply src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:111]: Cauuse tutors use the same endpoint to update proposed faq's as well

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (cremertim) [reply src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:78]: I used the same endpoints for create and update, cause otherwise i simply duplicate the code but only change the access check with @EnforceAtleastInstructor / @EnforeAtleastEditor and the name of the endpoint create/propose or update/updateProposed

File: src/main/webapp/i18n/de/faq.json
Comment (cremertim) [reply src/main/webapp/i18n/de/faq.json]: added title

File: src/main/webapp/app/faq/faq.component.ts
Comment (florian-glombik) [inline src/main/webapp/app/faq/faq.component.ts:50]: could be protected aswell

**Review by florian-glombik — DISMISSED** [review]
Approve code, thanks for addressing the comments and using the title in the alert instead of the id!

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by HawKhiem — COMMENTED** [review]
Tested on TS3. Tutor can't seem to post a FAQ, I sent details in the DM in Slack

**Review by HawKhiem — DISMISSED** [review]
Tested again on TS3. Everything works as described

Screenshot 2024-10-19 162114
Screenshot 2024-10-19 162133
Screenshot 2024-10-19 162140

**Review by florian-glombik — CHANGES_REQUESTED** [review]
The failing E2E tests seem to be related, could you fix them or verify that they are not related @cremertim ?

File: src/main/webapp/app/faq/faq.component.ts
Comment (florian-glombik) [inline src/main/webapp/app/faq/faq.component.ts:92]: You do not need to dismiss reviews for this little change, this would be a simpler way to express the same thing
```suggestion
         this.routeDataSubscription?.unsubscribe();
```

File: src/main/webapp/app/faq/faq.component.ts
Comment (florian-glombik) [inline src/main/webapp/app/faq/faq.component.ts:38]: I haven't noticed on the first review, in the future please always introduce signals. For this little change I do not mind

**Review by florian-glombik — DISMISSED** [review]
The code looks fine to me, I left 2 small suggestions that from my point of view to not need to keep the PR from being merged, rather little suggestions for future PRs :)

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (SimonEntholzer) [inline src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:195]: Is there any particular reason you use Set instead of List here?

**Review by SimonEntholzer — APPROVED** [review]
Code looks good, left one comment

File: src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java
Comment (cremertim) [reply src/main/java/de/tum/cit/aet/artemis/communication/web/FaqResource.java:195]: Not really. Intuitivly, a list might be more feasible, but should not behave differently cause two faqs cant be equal (due to the id). Do you wish that i chage it?

**Review by ahbitaqu — APPROVED** [review]
Tested on test-server3. Tried to break the feature with students role and tutor roles in addition to recommended instructions, but couldn't find any issues. Works as described & expected

**Review by flbrgit — COMMENTED** [review]
Tested on TS2 and all mechanics work as described:
Create _ Practical Course_ Interactive Learning WS24_25 - Vivaldi 20 10 2024 12_40_25
Create _ Practical Course_ Interactive Learning WS24_25 - Vivaldi 20 10 2024 12_42_32

There is a small graphics bug after proposing a new faq in the popup, where the {{ id }} is not rendered:
Create _ Practical Course_ Interactive Learning WS24_25 - Vivaldi 20 10 2024 12_40_25

**Review by flbrgit — CHANGES_REQUESTED** [review]
There's a small graphics bug in the popup after successfully creating a new FAQ as a tutor:
Create _ Practical Course_ Interactive Learning WS24_25 - Vivaldi 20 10 2024 12_40_25

How to reproduce: Follow the provided steps in the description of this PR and create a new FAQ. The bug occurs in the green success-popup

**Review by SindiBuklaji — APPROVED** [review]
Tested on TS3. The changes work as expected.👍

Small reminder for future edits: 
The cursor keeps blinking on this line while "+Category" remains written there. Otherwise great work! 
image

**Review by Malekos74 — APPROVED** [review]
### Server tested on: Server 1
It works as expected.
I like this feature :)
I also like the fact that the categories are not case sensitive (If I add the category FAQ and faq to two different questions, then they have the same color)

**Problems:**
No problems noticed.

**Minor notice:**
I noticed that the ID of the FAQ randomly started at 15 for some reason even though the FAQ ID's are only relevant for a certain course (So it should start at 1)
image

**Requests:**

- I think one should add more than 2 Categories, maybe cap it at 5.
- Maybe also have a course specific category list so that you don't have to type it each time.

**Review by janthoXO — APPROVED** [review]
Tested on TS1 and looks good. 👍

**Review by sarpsahinalp — APPROVED** [review]
Code lgtm

**Review by florian-glombik — APPROVED** [review]
Approve code, thanks for the adjustments and fixing the client tests!

~I did not double check the E2E tests, since Bamboo was not reachable for some reason. However, the last time I double checked the failing tests, the ones that were failing on this PR failed on develop.~ failing E2E tests are failing on develop aswell

**Review by zagemello — COMMENTED** [review]
Worked all as expected, tested on TS5 and rejected FAQs where not shown, whereas accepted ones appeared in the course.

**Review by ole-ve — APPROVED** [review]
code looks good after last approval

**Pull request merged**
